src/core/symbology/qgslinearreferencingsymbollayer.cpp

```diff
@@ -0,0 +1,1003 @@
+{
+  if ( !line )
+    return;
+
+  switch ( mPlacement )
+  {
+    case Qgis::LinearReferencingPlacement::IntervalCartesian2D:
+    case Qgis::LinearReferencingPlacement::IntervalZ:
+    case Qgis::LinearReferencingPlacement::IntervalM:
+      renderPolylineInterval( line, context, skipMultiples, QPointF( labelOffsetPainterUnitsX, labelOffsetPainterUnitsY ), averageAngleDistancePainterUnits, showMarker );
+      break;
+
+    case Qgis::LinearReferencingPlacement::Vertex:
+      renderPolylineVertex( line, context, skipMultiples, QPointF( labelOffsetPainterUnitsX, labelOffsetPainterUnitsY ), averageAngleDistancePainterUnits, showMarker );
+      break;
+  }
+}
+
+void QgsLinearReferencingSymbolLayer::renderPolyline( const QPolygonF &points, QgsSymbolRenderContext &context )
+{
+  QPainter *p = context.renderContext().painter();
+  if ( !p )
+  {
+    return;
+  }
+
+  double skipMultiples = mSkipMultiplesOf;
+  if ( mDataDefinedProperties.isActive( QgsSymbolLayer::Property::SkipMultiples ) )
+  {
+    context.setOriginalValueVariable( mSkipMultiplesOf );
+    skipMultiples = mDataDefinedProperties.valueAsDouble( QgsSymbolLayer::Property::SkipMultiples, context.renderContext().expressionContext(), mSkipMultiplesOf );
+  }
+
+  double labelOffsetX = mLabelOffset.x();
+  double labelOffsetY = mLabelOffset.y();
+
+  double averageOver = mAverageAngleLength;
+  if ( mDataDefinedProperties.isActive( QgsSymbolLayer::Property::AverageAngleLength ) )
+  {
+    context.setOriginalValueVariable( mAverageAngleLength );
+    averageOver = mDataDefinedProperties.valueAsDouble( QgsSymbolLayer::Property::AverageAngleLength, context.renderContext().expressionContext(), mAverageAngleLength );
+  }
+
+  bool showMarker = mShowMarker;
+  if ( mDataDefinedProperties.isActive( QgsSymbolLayer::Property::ShowMarker ) )
+  {
+    context.setOriginalValueVariable( showMarker );
+    showMarker = mDataDefinedProperties.valueAsBool( QgsSymbolLayer::Property::ShowMarker, context.renderContext().expressionContext(), mShowMarker );
+  }
+
+  const double labelOffsetPainterUnitsX = context.renderContext().convertToPainterUnits( labelOffsetX, mLabelOffsetUnit, mLabelOffsetMapUnitScale );
+  const double labelOffsetPainterUnitsY = context.renderContext().convertToPainterUnits( labelOffsetY, mLabelOffsetUnit, mLabelOffsetMapUnitScale );
+  const double averageAngleDistancePainterUnits = context.renderContext().convertToPainterUnits( averageOver, mAverageAngleLengthUnit, mAverageAngleLengthMapUnitScale ) / 2;
+
+  // TODO (maybe?): if we don't have an original geometry, convert points to linestring and scale distance to painter units?
+  // in reality this line type makes no sense for rendering non-real feature geometries...
+  ( void )points;
+  const QgsAbstractGeometry *geometry = context.renderContext().geometry();
+  if ( !geometry )
+    return;
```

Comment:
Should we not return earlier to save cost of computing nbs above if we're not going to render anything?

Reply:
Good point!